cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md

```diff
@@ -346,6 +346,63 @@ the device view can be obtained via function `column_device_view::create(column_
 `lists_column_device_view(column_device_view)`.
 
+# libcudf Policies and Design Principles
+
+`libcudf` is designed to provide GPU-accelerated algorithm primitives for solving a wide variety of problems that arise in data science.
+Our goal is to enable diverse use cases like Spark or Pandas to benefit from the performance of GPUs. 
+As a result, libcudf prioritizes performance and flexibility, which sometimes may come at the cost of convenience.
+While we welcome users to use libcudf directly, we design with the expectation that most users will be consuming libcudf through higher-level layers like Spark or cuDF Python that handle some of details that direct users of libcudf must handle on their own.
+We document these policies and the reasons behind them here:
+
+## libcudf does not introspect data
+
+libcudf APIs generally do not perform deep introspection and validation of input data.
+There are numerous reasons for this:
+1. It violates the single responsibility principle: validation is separate from execution.
+2. Since libcudf data structures store data on the GPU, any validation incurs _at minimum_ the overhead of a kernel launch, and may in general be prohibitively expensive.
+3. API promises around data introspection often significantly complicate implementation.
+
+Users are therefore responsible for passing valid data into such APIs.
+_Note that this policy does not mean that libcudf performs no validation whatsoever_.
+libcudf APIs should still perform any validation that does not require introspection.
+To give some idea of what should or should not be validated, here are (non-exhaustive) lists of examples.
+
+**Things that libcudf should validate**:
+- Input column/table sizes or dtypes
+
+**Things that libcudf should not validate**:
+- Integer overflow
+- Ensuring that outputs will not exceed the 2GB size limit for a given set of inputs
+
+
+## libcudf expects nested types to have sanitized null masks
+
+Various libcudf APIs accepting columns of nested dtypes (such as `LIST` or `STRUCT`) may assume that these columns have been sanitized.
+In this context, sanitization refers to ensuring that the null elements in a column with a nested dtype are compatible with the elements of nested columns.
+Specifically:
+- Null elements of list columns should also be empty. The starting offset of a null element should be equal to the ending offset.
+- Null elements of struct columns should also be null elements in the underlying structs.
+- For compound columns, nulls should only be present at the level of the parent column. Child columns should not contain nulls.
+- Slice operations on nested columns do not propagate offsets to child columns.
+
+**TODO: Do we still need to ensure the above with the new nested comparators? Since the nested comparators no longer operate based on flattening struct columns, I suspect that we will no longer need to superimpose parent nulls. We'll need to check that, though.**
```

Comment:
Maybe we can resolve this TODO before merging. I don't know exactly what is included in "ensure the above" but in my understanding we want to set the expectation that any/all APIs can _expect_ sanitized inputs with respect to nested nulls, whether or not they currently _rely_ on that behavior.

Reply:
Ah thanks for spotting that we hadn't addressed this yet. Yeah, that's all that I wanted to get across; even if the comparators don't require it (and as I stated above I think we can get away with it in some cases) we should set that as our API expectation. I'll remove this now.